### ISSUE TITLE
Issue with MatchesSMARTS when dealing with cyclic/symmetric molecules

### ISSUE THREAD

**Issue by alexandrefassio** [openbabel/openbabel#2131]
- [x] I believe this to be a bug with Open Babel
- [ ] This is a feature request

## Environment Information
Open Babel version: 3.0.0
Operating system and version: Linux

I found a possible issue with the function MatchesSMARTS() available at OBAtom when dealing with cyclic/symmetric molecules.

A simple example is the molecule "NC(=O)CCC1O[Mg]O1".

If I search for a pattern like "[OX2][C;X4,X3][OX2]" the result I got was False, but of course the pattern exists.

Then, I took a look at the file atom.cpp and I observed that MatchesSMARTS() is using the function GetUMapList() from OBSmartsPattern. The problem here is that GetUMapList() returns only unique matched patterns, so when we have cyclic/symmetric molecules, not all patterns are returned, but only one of them. As a consequence, the first atom of the unique list may not correspond to the target atom, so the pattern above works only for one of the Oxygens.

To solve this problem, I made my own function to match patterns by using GetMapList(). So, it returns all possible patterns, which means I can find all oxygens.

Therefore, I think you guys may update the function MatchesSMARTS by exchanging GetUMapList by GetMapList.

bests

**Comment by baoilleach** [comment]
Good catch. But if you're right, then this makes me shake my head. OBAtom::MatchesSmarts() should be using the specified OBAtom as the seed for the match. It shouldn't be running all over the molecule finding matches (unique or not ) and then asking whether any of the matches involved the specified atom. In the short term, we'll make this change but hopefully we can do this more efficiently in future.
